### PR TITLE
[Validator] check for empty host when calling checkdnsrr()

### PR DESCRIPTION
**Pull request by apetitpa**
| Q | A |
| --- | --- |
| Branch? | master|
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #22106 |
| License | MIT |
| Doc PR | n/a |

This should fix the email validator issue

### CODE REVIEW

**Comment by apetitpa** [comment]
Sorry for the last commit, I rebased branch with master and didn't notice you already did.

**Comment by apetitpa** [comment]
The hhvm build failed but it does not seem to be related to this PR.

**Comment by fabpot** [comment]
The regex should not be changed. But the condition should indeed be added in other methods using checkdnsrr.

**Comment by apetitpa** [comment]
thank you @fabpot and @jordscream I'll make some modifications, add unit tests and rebase my PR wwith 2.7

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php [hunk 1]
```diff
@@ -118,6 +118,10 @@ public function validate($value, Constraint $constraint)
      */
     private function checkMX($host)
     {
+        if (null === $host || '' === $host) {
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: `$host` can never be `null`, can it?

**Review by xabbuh — CHANGES_REQUESTED** [review]
Can you please also add a test to prevent future regressions?

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php (hunk 1 above)
Comment (apetitpa) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: I think you're right, I'll fix this. Thank you

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php [hunk 2]
```diff
@@ -118,6 +118,10 @@ public function validate($value, Constraint $constraint)
      */
     private function checkMX($host)
     {
+        if ('' === $host) {
+            return false;
+        }
+
```
Comment (jordscream) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: preferred empty host condition in return statement

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php [hunk 3]
```diff
@@ -130,6 +134,10 @@ private function checkMX($host)
      */
     private function checkHost($host)
     {
+        if ('' === $host) {
+            return false;
+        }
+
         return $this->checkMX($host) || (checkdnsrr($host, 'A') || checkdnsrr($host, 'AAAA'));
```
Comment (jordscream) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: preferred empty host condition in return statement

**Review by jsamouh — COMMENTED** [review]
What about unit test for the main issue ? ^^

**Comment by apetitpa** [comment]
About the unit test for the main issue, since checkMX() and checkHost() are both private how can I test it ?

**Comment by jordscream** [comment]
@apetitpa just test a use case  as "test@test@" or "test@" in the EmailValidator. It should return the appropriate violation

**Comment by apetitpa** [comment]
Thank you @jordscream I forgot that the original issue was with emails containing @ twice

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php [hunk 4]
```diff
@@ -118,7 +118,7 @@ public function validate($value, Constraint $constraint)
      */
     private function checkMX($host)
     {
-        return checkdnsrr($host, 'MX');
+        return ('' !== $host) && checkdnsrr($host, 'MX');
```
Comment (Nek-) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: why not `empty` check?

Comment (apetitpa) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: Well there is a similar check line 45 `if (null === $value || '' === $value)` that does not use the native empty function so I assumed I should do the same but maybe I was wrong ?

Comment (fabpot) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: We try to avoid using `empty` in Symfony.

Comment (fabpot) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: parentheses can be removed

Comment (apetitpa) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: I'll fix this in a minute thank you.

**Comment by jordscream** [comment]
@apetitpa , it seems your unit test fails cause you do not check the appropriate violation

**Comment by apetitpa** [comment]
@jordscream I think you're right I should check for MX_CHECK_FAILED_ERROR or HOST_CHECK_FAILED_ERROR

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php (hunk 4 above)
Comment (Nek-) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: @fabpot interesting, why? (is the reason documented somewhere?)

Comment (fabpot) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: `empty` is just too broad. For instance, it includes the `0` string as being `false`. Sometimes, that not desirable. So, we prefer to be very explicit about our expectations.

File: src/Symfony/Component/Validator/Constraints/UrlValidator.php [hunk 5]
```diff
@@ -72,7 +72,7 @@ public function validate($value, Constraint $constraint)
         if ($constraint->checkDNS) {
             $host = parse_url($value, PHP_URL_HOST);
 
-            if (!checkdnsrr($host, 'ANY')) {
+            if ('' === $host || !checkdnsrr($host, 'ANY')) {
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/UrlValidator.php]: Perhaps switch to `!is_string($host)` due possible `false` or `null` return values.

**Comment by jordscream** [comment]
@apetitpa . When you think to do it ? ^^. Need this fix for a customer :-P

File: src/Symfony/Component/Validator/Constraints/UrlValidator.php (hunk 5 above)
Comment (apetitpa) [inline src/Symfony/Component/Validator/Constraints/UrlValidator.php]: I think this is a good idea. I'll do it thank you.

**Comment by apetitpa** [comment]
@jordscream I hope this week-end but to be honest I'm facing difficulties creating an effective and working test for this issue. Some help would be appreciated.

**Comment by xabbuh** [comment]
> Dont think it needs to be dns-senstive though, as there will be no network traffic for `foo@bar@`.

Agreed, the issue to fix is that the DNS functions should in fact not be called anymore.

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php [hunk 6]
```diff
@@ -118,7 +118,7 @@ public function validate($value, Constraint $constraint)
      */
     private function checkMX($host)
     {
-        return checkdnsrr($host, 'MX');
+        return !empty($host) && checkdnsrr($host, 'MX');
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: Please avoid `empty` :) We expect a string, and only a empty string (`$host === ''`)  should not be passed to `checkdnsrr`.

File: src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php [hunk 7]
```diff
@@ -229,4 +229,32 @@ public function testHostnameIsProperlyParsed()
 
         $this->assertNoViolation();
     }
+
+    public function getCheckTypes()
```
Comment (ro0NL) [inline src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php]: Should be `provideCheckTypes` i guess, but it's a tiny semantic ;)

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php (hunk 6 above)
Comment (apetitpa) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: I know but `substr($value, strrpos($value, '@') + 1)` with `$value = foo@bar.fr@` returns `""` with PHP 7.* and `false` with PHP 5.6. So only checking for string emptiness is not enough.

File: src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php (hunk 7 above)
Comment (apetitpa) [inline src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php]: I'll make the modification thank you

File: src/Symfony/Component/Validator/Constraints/EmailValidator.php (hunk 6 above)
Comment (ro0NL) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: Let's cast the substring result above (or perhaps explicitly check for false); as currently we literally get `string('false')` as a replacement for `{{ value }}` in the violations above; due https://github.com/apetitpa/symfony/blob/master/src/Symfony/Component/Validator/ConstraintValidator.php#L139

So we got an extra (tiny) bug here ;-)

Comment (apetitpa) [inline src/Symfony/Component/Validator/Constraints/EmailValidator.php]: It's done, thank you

**Comment by apetitpa** [comment]
@xabbuh is it ok now ?

File: src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php [hunk 8]
```diff
@@ -229,4 +229,32 @@ public function testHostnameIsProperlyParsed()
 
         $this->assertNoViolation();
     }
+
+    public function provideCheckTypes()
```
Comment (fabpot) [inline src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php]: Can you move the provider after the test (that's the convention we are using in Symfony).

Comment (apetitpa) [inline src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php]: Of course, will do in a minute.

**Review by xabbuh — APPROVED** [review]
👍

**Comment by fabpot** [comment]
Thank you @apetitpa.

**Pull request closed without merging**